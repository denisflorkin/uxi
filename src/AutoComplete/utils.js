/*
Q: fernd
A:
[
  { matches: true, string: 'Fern' }, // !! watch out for the caps
  { matches: false, string: '&nbsp' },
  { matches: true, string: 'D' }, // !! watch out for the caps
  { matches: false, string: 'awson' },
]
*/

export const getMatchesResult = (source, target) => {
  // const source = 'Fernando Dawson'
  // const target = 'fernd'

  // shorcut in case of perfect match
  if (source.indexOf(target) === 0) {
    // ! perfect match, from the start worth a 100
    return [
      { matches: true, string: source.slice(0, target.length) },
      { matches: false, string: source.slice(target.length) },
    ]
  }

  const result = [];


  // this can probalby be refactored into a nice looking map + reducer:
  let i = 0;
  let j = 0;
  while (i < source.length) {
    const sourceChar = source[i];
    const targetChar = target[j];

    let currentRunMatchObj = {
      matches: false,
      string: sourceChar,
    };

    // console.log('sourceChar === targetChar')
    // console.log(`${sourceChar} === ${targetChar} : ${sourceChar === targetChar}`)

    if (sourceChar.toLowerCase() === targetChar && targetChar.toLowerCase()) {
      currentRunMatchObj.matches = true;
      // currentRunMatchObj.string += sourceChar
      result.push(currentRunMatchObj);
      // result.push(currentRunMatchObj)
      j++;
    } else {
      // currentRunMatchObj.string += sourceChar
      result.push(currentRunMatchObj);
      currentRunMatchObj = {
        matches: false,
        string: sourceChar,
      };
    }
    i++;
  }

  const initialAccu = {
    isCurrentlyMatching: result[0].matches,
    result: [{
      matches: result[0].matches,
      string: result[0].string,
    }],
  };

  const finalResult = result.reduce((accu, x, i) => {
    if (i === 0) { return accu }

    const currentItem = accu.result[accu.result.length - 1];

    if (x.matches === accu.isCurrentlyMatching) {
      accu.result[accu.result.length - 1].string += x.string;
      return accu;
    }
    return {
      isCurrentlyMatching: x.matches,
      result: [
        ...accu.result,
        { ...x },
      ],
    };
  }, initialAccu);

  // console.log('finalResult.result', finalResult.result)

  // add match score for later sorting
/*
  const finalResultWithScore = finalResult.reduce((finalResult, match) => {
    console.log('match', match);

    // const score = match.reduce((accu, x) => {
    //   // console.log('A: x.length', x.length)
    //   // console.log('A: x', x)

    //   if (a.matches) {
    //     return x.length > accu ? x.length : accu;
    //   }
    //   return accu
    // }, 0)

    return {
      ...match,
      // score,
    }
  }, finalResult)
  */
  // const sortedFinalResult = finalResult.result.sort((x, y) => {
  //   if (x.string.length > y.string.length) { return -1 }
  //   if (x.string.length < y.string.length) { return 1 }
  //   return 0
  // })

  /*
  for (let i = 0; i < source.length; i++) {
    const sourceChar = source[i];

    for (let j = 0; j < target.length; j++) {
      const targetChar = target[j];
      if (sourceChar === targetChar) {
        currentRunMatchObj.matches = true
        currentRunMatchObj.string += sourceChar
        // result.push(currentRunMatchObj)
      } else {
        // currentRunMatchObj.string += sourceChar
        result.push(currentRunMatchObj)
        // continue;
      }
    }
  }
  */
  // result
  // console.log(JSON.stringify(result, 2, 2))
  // console.log(JSON.stringify(finalResult.result, 2, 2))

  // return finalResultWithScore
  return finalResult.result
}


const addScore = (accu, { string, matches }) => {
  if (matches) {
    return string.length > accu ? string.length : accu;
  }
  return accu
}

export function getFilteredSetWithScore(filteredSet) {
  // console.log('filteredSet in sortFinaleResult', filteredSet)

  const filteredSetWithScore = filteredSet.map(x => ({
    matchList: x,
    scrore: x.reduce(addScore, 0),
  }))

  // console.log(filteredSetWithScore)

  return filteredSetWithScore;
}
