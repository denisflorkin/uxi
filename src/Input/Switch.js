import React, { PureComponent } from 'react';
import styled from 'styled-components';

const LabelWrapper = styled.label`
  display: flex;
  align-items: center;
  user-select: none;
  cursor: pointer;
  flex-direction: ${({ labelBefore }) => (labelBefore ? 'row-reverse' : 'row')};
`;

const SwitchOutterWrapper = styled.div`
  width: 42px;
  height: 20px;
  border-radius: 11px;
  border: 1px solid transparent;
  position: relative;
  margin: 0 6px 0 0;
  background: #cecece;
  background: ${({ theme, checked }) => (!checked ? theme.palette.lightGray : theme.palette.accent.main)};
`;

const SwitchInnerWrapper = styled.div`
  width: 16px;
  height: 16px;
  border-radius: 9px;
  border: 1px solid transparent;
  position: absolute;
  margin: 1px;
  transition: right .3s ease-out;
  background: white;
  left: ${({ checked }) => (!checked ? 0 : 'auto')};
  right: ${({ checked }) => (!checked ? 'auto' : 0)};
`;


const styles = {
  outterWrapper: {
    display: 'inline-block',
  },
  input: {
    display: 'none',
  },
  label: {
    margin: '0 6px 0 0',
  },
};

class Switch extends PureComponent {
  constructor(props) {
    super(props);
    this.state = {
      checked: false,
    };
    this.handleChange = this.handleChange.bind(this);
  }

  componentWillMount() {
    this.isControlled = this.props.checked !== undefined;
    if (!this.isControlled) {
      // not controlled, use internal state
      this.setState({
        checked: this.props.defaultChecked !== undefined ? this.props.defaultChecked : false,
      });
    }
  }

  handleChange(event) {
    const { checked } = this.state;
    const { onChange } = this.props;
    const newState = (!checked);
    if (!this.isControlled) {
      this.setState({
        checked: newState,
      });
    }
    if (onChange) { onChange({ checked: newState }, event); }
  }

  render() {
    const { label, name, id, labelBefore, checked } = this.props;

    const checker = this.isControlled ? checked : this.state.checked;

    return (
      <div style={styles.outterWrapper} >
        <LabelWrapper
          labelBefore={labelBefore}
          htmlFor={name || 'myAwesomeCheckBox'}
          onClick={this.handleChange}
        >
          <SwitchOutterWrapper checked={checker}>
            <SwitchInnerWrapper checked={checker} />
          </SwitchOutterWrapper>
          <div style={styles.label}>
            {label}
          </div>
        </LabelWrapper>
        <input id={name || 'myAwesomeCheckBox'} style={styles.input} type="checkbox" />
      </div>
    );
  }
}

export default Switch;
