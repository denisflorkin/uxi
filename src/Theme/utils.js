
import { palette } from './palette';

// eslint-disable-next-line no-shadow
function mergeCustomPalette(mainPalette, customPalette) {
  let primary = mainPalette.primary.main;
  let secondary = mainPalette.accent.main;

  if (!customPalette) { return mainPalette; }
  if (customPalette.primary) { primary = customPalette.primary; }
  if (customPalette.secondary) { secondary = customPalette.secondary; }

  const palette = { // eslint-disable-line no-shadow
    ...mainPalette,
    primary: {
      main: primary,
      light: lighten(primary),
      dark: darken(primary),
    },
    accent: {
      main: secondary,
      light: lighten(secondary),
      dark: darken(secondary),
    },
  };
  return palette;
}

// eslint-disable-next-line no-use-before-define
export const getThemeWithCustomPalette = (customPalette = {}) => { // eslint-disable-line no-shadow
  return {
    ...theme,
    palette: mergeCustomPalette(palette, customPalette),
  };
};
