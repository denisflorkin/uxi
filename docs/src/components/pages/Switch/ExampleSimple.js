import React from 'react';
import { Checkbox } from '../../../../../src/Input';

const ExampleSimple = () => (
  <ul>
    <li>
      <h3>Checkbox :</h3>
      <Checkbox />
    </li>
    <li>
      <h3>Checkbox disabled</h3>
      <Checkbox disabled />
    </li>
  </ul>
);

export default ExampleSimple;
