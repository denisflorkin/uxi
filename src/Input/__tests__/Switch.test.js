import Switch from '../Switch';
import React from 'react';
import { mountWithTheme } from '../../../test/utils';

const defaultProps = { };

const getCompo = props => (
  shallow(
    // mountWithTheme(
    <Switch
      {...defaultProps}
      {...props}
    />
    // )
  )
)

describe('<Switch />', () => {
  it('exists', () => {
    expect(Switch).toBeDefined()
  })

  it('matches snapshot', () => {
    const wrapper = getCompo();
    expect(wrapper).toMatchSnapshot();
  })

  describe('Uncontrolled', () => {
    it('handles its own state', () => {
      const wrapper = getCompo({});

      expect(wrapper.instance().isControlled).toBe(false);
      expect(wrapper.state().checked).toBe(false);
      expect(wrapper.props().checked).toBeUndefined();
    })
  })

  describe('Controlled', () => {
    it('respect the props passed down no matter what', () => {
      const wrapper = shallow(<Switch checked />);
      console.log(wrapper)
      console.log(wrapper.props().checked)

      expect(wrapper.instance().isControlled).toBe(true);
      // expect(wrapper.props().checked).toBe(true);
      // expect(wrapper.props().checked).toBe(true);
    })
  })
})
