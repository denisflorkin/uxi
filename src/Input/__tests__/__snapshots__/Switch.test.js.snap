// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Switch /> matches snapshot 1`] = `
<div
  style={
    Object {
      "display": "inline-block",
    }
  }
>
  <styled.label
    htmlFor="myAwesomeCheckBox"
  >
    <styled.div
      checked={false}
    >
      <styled.div
        checked={false}
      />
    </styled.div>
    <styled.div />
    <styled.input
      name={0}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      style={Object {}}
      type="checkbox"
    />
  </styled.label>
</div>
`;
